js/trackers.js

```diff
@@ -233,8 +231,48 @@ require.scopes.trackers = (function() {
         return match
     }
 
+    // add block/noblock decisions (per domain) to the trackers cache
+    // remove older entries if the cache is full
+    function addToCache (reqUrl, currLocation, trackerData) {
+        if (!settings.getSetting('trackerBlockingEnabled')) return
+        if (isFirstPartyRequest(currLocation, reqUrl)) return
+        if (cache.size > 3000) {
+            cache.delete(cache.keys().next().value)
```

Comment:
I'm not sure about the implementation but this looks super inefficient. Like, you'll reach 3000 and then this will be called for every single URL from then on out.

Reply:
I thought about this, too, at the outset of this task and wrote a quick script you can copy/paste into the browser and futz around with. It's a timer I wrote to make sure that `Map` would be just as efficient as a POJO (plain old javascript object) and it indeed it seemed to be. You'll notice that the very first key added and retrieved from the cache (WITHOUT the delete operation) takes ~2ms total. From the iteration no. 2 - 2999, on Chrome it gets incredibly efficient, the whole operation (without the delete) takes ~0.04-0.06ms. Once the delete operation is added to the mix after iteration 3000, on my machine all of those operations -- both adding the cache key and then retrieving it -- take ~0.25ms total. 

Firefox's performance is slighly slower. Again, the very first key added to the Map and its retrieval takes about 2ms, and from iteration 2-2999, im seeing it take about ~0.2 - 0.6ms. Once the delete operation is added to the mix after iteration 3000, im seeing it take ~0.4 - 0.8ms. 

This seems pretty efficient to me, and much faster than the raw tracker lookups. The script that I used is below, feel free to copy/paste into the console of any browser and see how it looks to you if you like. Am open to any further comments/observations, but bear in mind that I took this into consideration quite a bit.

var cache = new Map()

function addToCache (key, cancel) {
    if (cache.size >= 3000) {
        console.log('DELETE')
        cache.delete(cache.keys().next().value)
    }
    cache.set(key, cancel)
}

function isCached (url) {
    return new Promise((resolve, reject) => {
        var cachedValue = cache.get(url)
        resolve({cancel: cachedValue})
    })
}

for (var i=0; i <= 3500; i++) {
    ms = 50
    counter = 0
    setTimeout(function () {
        counter++
        var r = () => (Math.random() * 1000000000000000000000).toString()
        // key length is 64 characters, same length as a SHA256 hash
        var key = r() + r() + r() + 'f'

        // time these operations, output to console
        console.time(counter)
        addToCache(key, true)
        isCached(key).then((v) => true )
        console.timeEnd(counter)

    }, ms)
}

Reply:
Thinking about it more -- one improvement that could be made here: the .addToCache() step can be made to be non-blocking -- meaning nothing that happens in the `onBeforeRequest` handler after .addToCache() is called is dependent on it having run -- so I can actually speed this up even further by returning a promise immediately from .addToCache() -- even though we won't do anything with the promise(!), we'll just use it to make .addToCache() return immediately, which in turn means that the `onBeforeRequest()` handler will return faster, too. https://github.com/duckduckgo/chrome-zeroclickinfo/pull/231/files#diff-2650fb3e14b7f42d68a8766269c54434R141

Reply:
good to know.

this 3000 spans all tabs right? what about the idea of one cache per page? you could delete that whole thing when page is finished loading, which would also address your concern about storing history.

also - do we know the order of cache.keys()? will there be cases where, after 3000 we're not caching anything (or only partially) on the subsequent pages, for example, the 3001th is my first image url on amazon.com, and it so happens that the cache.keys().next() call returns this so nothing ends up getting cached on amazon?

Reply:
Actually I love the idea of caching per page instead of in one huge glob that spans all tabs like this. Am assuming you mean that we could eliminate the hashing operation if this were the case and we cached by page (or tab). That would be the superior design for sure(!). All along I've been weighing the performance pickup we'd get on sites with lots of tracker lookups vs. the drag of the caching operation on pages with lots of 3rd party scripts that are not trackers. This also solves the history problem. 

(The second concern you raised above is not valid if I'm understanding it correctly, bc all the keys are hashed).